[performance] - Remove bulk and streams use cases from UO and TO. Scallability test added.

see-quick opened the pull request:
### Type of change

- Enhancement / new feature

### Description

This PR focuses on exploring the impact of different configurations on the efficiency of creating, modifying, and deleting Kafka topics. I've played around with a range of batch sizes and linger durations to see how they affect performance across different scales of topic counts.

Based on this graph (KRaft):

![image](https://github.com/strimzi/strimzi-kafka-operator/assets/30839163/d9d92d8a-c911-4557-9cf0-693b1ef15fd6)

One can see that I have tried multiple configurations with batch sizes and linger settings stretching from 1ms to 2000ms. Moreover, the range of topics which I tested was from 50 to 1000 to see some pattern if such configuration scaling well or if there are some problems....(that could be viewed on each curve). This could help us understand the capabilities of UTO with various settings and scale with the best configuration. 

I have also implemented the way how we create the events. Currently, we are doing it sequentially and now I have modified it and used ExecutorService to manage and process batches concurrently. More on that in the Javadoc...

[1] - https://github.com/strimzi/strimzi-kafka-operator/pull/10050#pullrequestreview-2039592438

Update (19.9.2024):

After a few modifications, also we have decided to remove two use cases from TO and UO (i.e., Alice bulk and bob's streaming). We do not think it adds much value so we would currently stick to capacity and scalability tests, which are now present in those test suites. 

### Checklist

- [x] Write tests
- [x] Make sure all tests pass

see-quick commented:
/packit test --labels performance-topic-operator-capacity

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testCapacityCreateAndUpdateTopics --env=STRIMZI_USE_KRAFT_IN_TESTS=true

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testCapacityCreateAndUpdateTopics --env=STRIMZI_USE_KRAFT_IN_TESTS=true

fvaleri reviewed (commented):
Hi @see-quick, thanks for working on this.

In order to simulate a busy shared cluster and possibly catch some edge cases, I think we should try to include all 3 kind of topic events (creations, updates and deletes) and run them in parallel.

In my custom test, I'm taking the number of events I want to test as input, then I divide them by 3 to get the number of tasks I have to run in parallel (you would have 1/2 spare events that you can simply consume as noop, that's fine). Each task executes topic creation, update (partition increase and config change) and deletion serially. Wdyt?

see-quick commented:
> Hi @see-quick, thanks for working on this.
> 
> In order to simulate a busy shared cluster and possibly catch some edge cases, I think we should try to include all 3 kind of topic events (creations, updates and deletes) and run them in parallel.
> 
> In my custom test, I'm taking the number of events I want to test as input, then I divide them by 3 to get the number of tasks I have to run in parallel (you would have 1/2 spare events that you can simply consume as noop, that's fine). Each task executes topic creation, update (partition increase and config change) and deletion serially. Wdyt?

Okay, but that way we would not be able to see the upper bound (i.e., how many KafkaTopics is TO able to handle in creation and modification). Maybe such information is not so important....and if we make all these three operations what is our termination condition? Do we want to create a specific number of topics (e.g., 1000) and see how TO is performing on different configurations? What are then the most important **OUT metrics** to check? Also should we execute these tasks incrementally and divide them into batches (i.e., every 100 KafkaTopics?) as we do capacity or should we run all  1000 topics in once?

fvaleri commented:
> we would not be able to see the upper bound (i.e., how many KafkaTopics is TO able to handle in creation and modification)

I think the objective here is not see the upper bound, but to assess performance in a fixed size of events. For example, I'm running test with the following batch of events: 50, 100, 150, ..., 1000. That way, you see how it scales, by simply putting the end-to-end reconciliation (we only care about this one here) time on a line graph, and you can compare with a previous implementation on the very same graph.

With e2e reconciliation time in seconds I mean the time from creation/update to ready, or deletion duration. This is how an example graph looks like (note: we only need number, then you can generate the graph with whatever tool you prefer):

![Screenshot from 2024-05-27 11-01-57](https://github.com/strimzi/strimzi-kafka-operator/assets/11456498/660cc026-4de7-4473-81d4-e69c221f6b9d)

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testPerformanceInFixedSizeOfEvents --env=STRIMZI_USE_KRAFT_IN_TESTS=true

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testPerformanceInFixedSizeOfEvents --env=STRIMZI_USE_KRAFT_IN_TESTS=true

see-quick commented on systemtest/src/main/java/io/strimzi/systemtest/Environment.java:
```diff
@@ -226,7 +226,7 @@ public class Environment {
     public static final boolean SKIP_TEARDOWN = getOrDefault(SKIP_TEARDOWN_ENV, Boolean::parseBoolean, false);
     public static final String STRIMZI_RBAC_SCOPE = getOrDefault(STRIMZI_RBAC_SCOPE_ENV, STRIMZI_RBAC_SCOPE_DEFAULT);
     public static final String STRIMZI_FEATURE_GATES = getOrDefault(STRIMZI_FEATURE_GATES_ENV, STRIMZI_FEATURE_GATES_DEFAULT);
-    public static final boolean STRIMZI_USE_KRAFT_IN_TESTS = getOrDefault(STRIMZI_USE_KRAFT_IN_TESTS_ENV, Boolean::parseBoolean, false);
```
I will revert this back, it was just for testing

see-quick commented:
> > we would not be able to see the upper bound (i.e., how many KafkaTopics is TO able to handle in creation and modification)
> 
> I think the objective here is not see the upper bound, but to assess performance in a fixed size of events. For example, I'm running test with the following batch of events: 50, 100, 150, ..., 1000. That way, you see how it scales, by simply putting the end-to-end reconciliation (we only care about this one here) time on a line graph, and you can compare with a previous implementation on the very same graph.
> 
> With e2e reconciliation time in seconds I mean the time from creation/update to ready, or deletion duration. This is how an example graph looks like (note: we only need number, then you can generate the graph with whatever tool you prefer):
> 
> ![Screenshot from 2024-05-27 11-01-57](https://private-user-images.githubusercontent.com/11456498/334036529-660cc026-4de7-4473-81d4-e69c221f6b9d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwNTUxNjYsIm5iZiI6MTcxNzA1NDg2NiwicGF0aCI6Ii8xMTQ1NjQ5OC8zMzQwMzY1MjktNjYwY2MwMjYtNGRlNy00NDczLTgxZDQtZTY5YzIyMWY2YjlkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMwVDA3NDEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA4MDc5MWI4MGFjYThiNjYyY2NjMmNlMjE2MzUyMzExYzgwN2VjNWFkNDA2OWEzNmZjZTRlZjg3NDhjYTZlZTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.B9m1GutFsH0Tcpv5iirsu2ZqLC8ZjCEKtDcYT7vNvr0)

So I have tried 6 configurations here.

![image](https://github.com/strimzi/strimzi-kafka-operator/assets/30839163/1cd0379d-29df-45b1-bfad-c319ef2414ac)

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -740,6 +743,136 @@ void testCapacity(String maxBatchSize, String maxBatchLingerMs) throws IOExcepti
+            resourceManager.deleteResourcesOfTypeWithoutWait(KafkaTopic.RESOURCE_KIND);
+            KafkaTopicUtils.waitForTopicWithPrefixDeletion(testStorage.getNamespaceName(), testStorage.getTopicName());
+            if (this.topicOperatorMetricsGatherer != null) {
+                this.topicOperatorMetricsGatherer.stopCollecting();
```
We should stop collecting metrics just after `processKafkaTopicBatchesAsync`. Deleting topics in the `finally` block is just a safe net in case something goes wrong and some KT are not deleted.

see-quick replied:
> Deleting topics in the finally block is just a safe net in case something goes wrong and some KT is not deleted.

Yeah, true I will re-order this two ;)

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,214 @@
+     * @param testStorage    The storage instance containing the configuration and namespace details.
+     * @param numberOfTopics The total number of Kafka topics to process.
+     */
+    public static void processKafkaTopicBatchesAsync(TestStorage testStorage, int numberOfTopics) {
```
Here you are batching in terms of topic resources. On the other end, the operator actually reconciles topic events, and this is what is measured by metrics.

A topic event has one of 2 types: `upsert` (create, update), `delete`. An update event can be further divided into 2 sub-types: `resync` (from periodic reconciliation), `update` (on-demand update).

Using `numberOfTopics` is fine, but you are actually generating a workload of `numberOfTopics * 3` events for the operator. This should be reflected on the X axis.

see-quick replied:
Okay so what about adding a a `IN` metric with `NUMBER_OF_EVENTS`, which would be equal to `numberOfTopics` * 3?

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,214 @@
+     * @param executor        The executor service managing concurrent tasks.
+     * @return                CompletableFuture representing the completion of all operations in the batch.
+     */
+    private static CompletableFuture<Void> processBatch(int start, int end, ExtensionContext currentContext,
```
The way you process the batch is not similar to what a normal user would do. He would create 1 topic, do some updates, and maybe delete it when done. Instead here we are creating, updating and deleting all at once.

see-quick replied:
Sure I have impl. such approach in next commit :) check it

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,214 @@
+        final CompletableFuture<?>[] futures = new CompletableFuture[optimalBatchCount];
+
+        try {
+            for (int batch = 0; batch < optimalBatchCount; batch++) {
```
Should we do some warmup before this line?

see-quick replied:
Does it really change anything?

fvaleri reviewed (commented):
@see-quick nice work.

I left some improvement suggestions, but the base logic is there.

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testPerformanceInFixedSizeOfEvents --env=STRIMZI_USE_KRAFT_IN_TESTS=true

see-quick commented:
/packit test --labels performance-topic-operator-user-workload

see-quick commented:
/packit test --labels performance-topic-operator-user-workload

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testPerformanceInFixedSizeOfEvents --env=STRIMZI_USE_KRAFT_IN_TESTS=true

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testPerformanceInFixedSizeOfEvents --env=STRIMZI_USE_KRAFT_IN_TESTS=true

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,272 @@
+     * @param testStorage       An instance of TestStorage containing configuration and state needed for topic operations.
+     * @param numberOfTopics    The number of Kafka topics to be processed.
+     */
+    public static void processAllTopicsConcurrently(TestStorage testStorage, int numberOfTopics) {
```
Should this be part of the test? It's part of the logic and it's only used there.

fvaleri replied:
That's premature optimization :)

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,272 @@
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
+        LOGGER.info("All topic lifecycles completed.");
+
+        executor.shutdown();
```
This may not be enough in some situations. Something like this is better:

```sh
        try {
            executor.shutdown();
            executor.awaitTermination(timeoutMs, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            if (!executor.isTerminated()) {
                executor.shutdownNow();
            }
        }
```

see-quick replied:
Sure :) I would add this thanks

fvaleri reviewed (commented):
When I run the new test I get:

```sh
[ERROR] io.strimzi.systemtest.performance.TopicOperatorPerformance.testPerformanceInFixedSizeOfEvents(String, String, boolean, int)[20] -- Time elapsed: 180.2 s <<< ERROR!
java.util.concurrent.CompletionException: java.lang.RuntimeException: Using type: null which is not supported!
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1807)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.RuntimeException: Using type: null which is not supported!
	at io.strimzi.systemtest.utils.StUtils.lambda$static$1(StUtils.java:79)
	at io.strimzi.systemtest.utils.StUtils.isParallelNamespaceTest(StUtils.java:404)
	at io.strimzi.systemtest.resources.ResourceManager.setNamespaceInResource(ResourceManager.java:256)
	at io.strimzi.systemtest.resources.ResourceManager.createResource(ResourceManager.java:181)
	at io.strimzi.systemtest.resources.ResourceManager.createResourceWithoutWait(ResourceManager.java:168)
	at io.strimzi.systemtest.utils.kafkaUtils.KafkaTopicScalabilityUtils.createTopicsViaK8s(KafkaTopicScalabilityUtils.java:48)
	at io.strimzi.systemtest.performance.utils.TopicOperatorPerformanceUtils.performCreationWithWait(TopicOperatorPerformanceUtils.java:128)
	at io.strimzi.systemtest.performance.utils.TopicOperatorPerformanceUtils.performFullLifecycle(TopicOperatorPerformanceUtils.java:268)
	at io.strimzi.systemtest.performance.utils.TopicOperatorPerformanceUtils.lambda$processAllTopicsConcurrently$8(TopicOperatorPerformanceUtils.java:247)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
```

see-quick commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -740,6 +743,143 @@ void testCapacity(String maxBatchSize, String maxBatchLingerMs) throws IOExcepti
+    @Tag(PERFORMANCE)
+    @ParameterizedTest
+    @MethodSource("provideConfigurationsFortestPerformanceInFixedSizeOfEvents")
+    void testPerformanceInFixedSizeOfEvents(String maxBatchSize, String maxBatchLingerMs, boolean processEachTopicSeparately, int numberOfTopics) throws IOException {
```
sure good idea also I would add a specific tag `@Tag(SCALABILITY)`

see-quick commented:
> When I run the new test I get:
> 
> ```shell
> [ERROR] io.strimzi.systemtest.performance.TopicOperatorPerformance.testPerformanceInFixedSizeOfEvents(String, String, boolean, int)[20] -- Time elapsed: 180.2 s <<< ERROR!
> java.util.concurrent.CompletionException: java.lang.RuntimeException: Using type: null which is not supported!
> 	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
> 	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1807)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
> 	at java.base/java.lang.Thread.run(Thread.java:840)
> Caused by: java.lang.RuntimeException: Using type: null which is not supported!
> 	at io.strimzi.systemtest.utils.StUtils.lambda$static$1(StUtils.java:79)
> 	at io.strimzi.systemtest.utils.StUtils.isParallelNamespaceTest(StUtils.java:404)
> 	at io.strimzi.systemtest.resources.ResourceManager.setNamespaceInResource(ResourceManager.java:256)
> 	at io.strimzi.systemtest.resources.ResourceManager.createResource(ResourceManager.java:181)
> 	at io.strimzi.systemtest.resources.ResourceManager.createResourceWithoutWait(ResourceManager.java:168)
> 	at io.strimzi.systemtest.utils.kafkaUtils.KafkaTopicScalabilityUtils.createTopicsViaK8s(KafkaTopicScalabilityUtils.java:48)
> 	at io.strimzi.systemtest.performance.utils.TopicOperatorPerformanceUtils.performCreationWithWait(TopicOperatorPerformanceUtils.java:128)
> 	at io.strimzi.systemtest.performance.utils.TopicOperatorPerformanceUtils.performFullLifecycle(TopicOperatorPerformanceUtils.java:268)
> 	at io.strimzi.systemtest.performance.utils.TopicOperatorPerformanceUtils.lambda$processAllTopicsConcurrently$8(TopicOperatorPerformanceUtils.java:247)
> 	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
> ```

Hmmm, this seems that ExtensionContext is not properly propagated into another Thread.... :) let me check that.

fvaleri commented:
Thanks for fixing the issue. I will try to run again with bs100,lm10 and compare with my test driver. Results should be close.

fvaleri commented:
It's kind hard to compare, because of the input mismatch (n.events vs n.topics), but the results I got seems strange (yellow line up to 300 events).

![Screenshot from 2024-06-05 17-39-14](https://github.com/strimzi/strimzi-kafka-operator/assets/11456498/62cc00c8-6770-423a-b62f-4677c5fd19da)

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testPerformanceInFixedSizeOfEvents --env=STRIMZI_USE_KRAFT_IN_TESTS=true

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,286 @@
+     * @param testStorage       An instance of TestStorage containing configuration and state needed for topic operations.
+     * @param numberOfTopics    The number of Kafka topics to be processed.
+     */
+    public static void processAllTopicsConcurrently(TestStorage testStorage, int numberOfTopics, int spareEvents) {
```
I guess it's numerOfEvents or events now. Warmup phase still missing.

see-quick replied:
No, this is still topics or numberOfTasks I don't know what you prefer... :) Let me know

fvaleri replied:
Right. I think numberOfTasks is better.

see-quick replied:
Updated

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,286 @@
+     */
+    public static void processAllTopicsConcurrently(TestStorage testStorage, int numberOfTopics, int spareEvents) {
+        final int availableCPUs = Math.max(1, Runtime.getRuntime().availableProcessors());
+        final ExecutorService executor = Executors.newFixedThreadPool(availableCPUs);
```
I would not include executor service init/shudown into the e2e reconciliation time.

see-quick replied:
done :)

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testSystemScalability --env=STRIMZI_USE_KRAFT_IN_TESTS=true

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testSystemScalability --env=STRIMZI_USE_KRAFT_IN_TESTS=true

see-quick commented:
> Now I can easily compare, and I see that the e2e reconciliation time from your test is roughly doubled compared to mine. Left some comments that may help.

Now you can try :) But I am afraid you would still get higher e2e RT from our tests. I see this 
```java
| Experiment | IN: MAX QUEUE SIZE | IN: MAX BATCH SIZE (ms) | IN: NUMBER OF TOPICS | IN: NUMBER OF EVENTS | IN: MAX BATCH LINGER (ms) | IN: PROCESS TYPE | OUT: Successful KafkaTopics Created and Modified and Deleted (ms) | jvm_memory_used_megabytes_total.txt | strimzi_update_status_duration_seconds_max.txt | strimzi_reconciliations_max_queue_size.txt | system_load_average_per_core.txt | strimzi_reconciliations_max_batch_size.txt | strimzi_reconciliations_duration_seconds_max.txt | strimzi_create_topics_duration_seconds_max.txt | strimzi_describe_configs_duration_seconds_max.txt | strimzi_total_time_spend_on_uto_event_queue_duration_seconds.txt |
| 1          | 2147483647         | 100                     | 16                   | 50                   | 100                       | TOPIC-CONCURRENT | 4697                                                              | 118.340976                          | 0.131897589                                    | 56.0                                       | 0.155                            | 98.0                                       | 0.847086864                                      | 0.124908912                                    | 0.030741532                                       | 151.562578987                                                    |
| 2          | 2147483647         | 100                     | 33                   | 100                  | 100                       | TOPIC-CONCURRENT | 10759                                                             | 128.704712                          | 0.245771336                                    | 104.0                                      | 0.08875                          | 98.0                                       | 0.540274112                                      | 0.083525793                                    | 0.043971874                                       | 143.946798066                                                    |
| 3          | 2147483647         | 100                     | 16                   | 50                   | 10                        | TOPIC-CONCURRENT | 4723                                                              | 172.27056                           | 0.049315916                                    | 21.0                                       | 0.24375                          | 100.0                                      | 0.475620992                                      | 0.085987189                                    | 0.022740989                                       | 124.457906132                                                    |
| 4          | 2147483647         | 100                     | 33                   | 100                  | 10                        | TOPIC-CONCURRENT | 10544                                                             | 130.051632                          | 0.040530541                                    | 66.0                                       | 0.13875                          | 99.0                                       | 0.554755079                                      | 0.085650638                                    | 0.028386005                                       | 163.853588678                                                    |
| 5          | 2147483647         | 10                      | 16                   | 50                   | 1                         | TOPIC-CONCURRENT | 4709                                                              | 116.326584                          | 0.028667492                                    | 93.0                                       | 0.13                             | 10.0                                       | 0.458799716                                      | 0.085642402                                    | 0.013513653                                       | 71.31604381700001                                                |
| 6          | 2147483647         | 10                      | 33                   | 100                  | 1                         | TOPIC-CONCURRENT | 10504                                                             | 127.834504                          | 0.08446808                                     | 91.0                                       | 0.2275                           | 10.0                                       | 0.409806775                                      | 0.078609904                                    | 0.028111449                                       | 79.053662579                                                     |
| 7          | 2147483647         | 50                      | 16                   | 50                   | 100                       | TOPIC-CONCURRENT | 4694                                                              | 115.719672                          | 0.154474669                                    | 117.0                                      | 0.12                             | 50.0                                       | 0.428688145                                      | 0.092419893                                    | 0.022318055                                       | 98.574897885                                                     |
| 8          | 2147483647         | 50                      | 33                   | 100                  | 100                       | TOPIC-CONCURRENT | 10501                                                             | 123.497136                          | 0.042973893                                    | 61.0                                       | 0.07375                          | 50.0                                       | 0.428938314                                      | 0.082578837                                    | 0.021978298                                       | 108.432313159                                                    |
| 9          | 2147483647         | 100                     | 16                   | 50                   | 500                       | TOPIC-CONCURRENT | 4602                                                              | 128.54072                           | 0.032225636                                    | 61.0                                       | 0.11375                          | 99.0                                       | 0.489799132                                      | 0.077477801                                    | 0.026273763                                       | 122.413640575                                                    |
| 10         | 2147483647         | 100                     | 33                   | 100                  | 500                       | TOPIC-CONCURRENT | 10447                                                             | 130.3082                            | 0.047551127                                    | 19.0                                       | 0.11375                          | 100.0                                      | 0.518047899                                      | 0.075547221                                    | 0.024660648                                       | 154.833472284                                                    |
| 11         | 2147483647         | 500                     | 16                   | 50                   | 1000                      | TOPIC-CONCURRENT | 4611                                                              | 128.366336                          | 0.059608899                                    | 104.0                                      | 0.25375                          | 98.0                                       | 0.499648783                                      | 0.243576891                                    | 0.021484391                                       | 134.952340021                                                    |
| 12         | 2147483647         | 500                     | 33                   | 100                  | 1000                      | TOPIC-CONCURRENT | 10434                                                             | 128.675792                          | 0.062352485                                    | 101.0                                      | 0.1025                           | 99.0                                       | 0.561572505                                      | 0.169018705                                    | 0.04502545                                        | 147.17968550900002                                               |
+------------+--------------------+-------------------------+----------------------+----------------------+---------------------------+------------------+-------------------------------------------------------------------+-------------------------------------+------------------------------------------------+--------------------------------------------+----------------------------------+--------------------------------------------+--------------------------------------------------+------------------------------------------------+---------------------------------------------------+------------------------------------------------------------------+
```

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,295 @@
+     */
+    public static void processKafkaTopicBatchesAsync(TestStorage testStorage, int numberOfTopics) {
+        final int availableCPUs = Runtime.getRuntime().availableProcessors();
+        final int optimalBatchCount = Math.max(1, availableCPUs);
```
availableProcessors is never smaller than one, so you can save some cycles here

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,295 @@
+     * @return                      The total time taken to complete all topic lifecycles in milliseconds.
+     */
+    public static long processAllTopicsConcurrently(TestStorage testStorage, int numberOfTopics, int spareEvents, int warmUpTasksToProcess) {
+        final int availableCPUs = Math.max(1, Runtime.getRuntime().availableProcessors());
```
availableProcessors is never smaller than one, so you can save some cycles here

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,295 @@
+     */
+    public static long processAllTopicsConcurrently(TestStorage testStorage, int numberOfTopics, int spareEvents, int warmUpTasksToProcess) {
+        final int availableCPUs = Math.max(1, Runtime.getRuntime().availableProcessors());
+        final ExecutorService executor = Executors.newFixedThreadPool(availableCPUs);
```
I also did this wrong, but we could actually initialize this once statically at the class level, and set up a shutdown hook to stop it.

see-quick replied:
Hmm...that way it would be better to have some abstraction....handling this as an object rather than a Util class... :) but I would rather leave for the next PR.

fvaleri replied:
IMO, it's a small pragmatic improvement, without a huge impact on test results.

see-quick replied:
Updated

fvaleri commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,295 @@
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
+        LOGGER.info("All topic lifecycles completed.");
+
+        long allTasksTimeMs = (System.nanoTime() - startTime) / 1_000_000;
```
It would be good to add some sleep time (e.g. 30s) as cooldown after this line (i.e after each batch completes).

see-quick replied:
Why it's needed?

fvaleri replied:
It's just a way to set a boundary between tests, so that one test run is less likely to influence results of the next.

see-quick replied:
Cool I have added
```java
if (warmUpTasksToProcess != 0) {
            // boundary between tests => less likelihood that tests would influence each other
            LOGGER.info("Cooling down");
            Thread.sleep(30_000);
        }
```

fvaleri reviewed (commented):
Results are much better now. Left some more suggestions :)

I also have a couple of questions:

- Do we want to switch to ZK based cluster for the time being, as this is what most users have?
- Are you reserving CPU/memory? We need some level of isolation to generate comparable results

see-quick commented:
> Do we want to switch to ZK based cluster for the time being, as this is what most users have?

I am not sure with all those ZK-DNS-related problems it might be a problem sometimes...where we deploy a lot of clusters and that could result in many flaky tests...and not-so-great results. I don't know, I would rather stick with KRaft....or at least we could once compare results ZK and KRaft-based....

> Are you reserving CPU/memory? We need some level of isolation to generate comparable results|

For each run, I am using same multi-node cluster, which is based on same flavour of machines :)

fvaleri commented:
> ZK-DNS-related problems it might be a problem sometimes

It always work fine for me on Minikube with 3 replicas. As I said, we should test against a deployment that most users care. We will switch to KRaft as soon as we see that most users are migrated, which they will be forced to do at some point.

> For each run, I am using same multi-node cluster, which is based on same flavour of machines :)

Ok, fine. Thanks.

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testSystemScalability

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testSystemScalability

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -737,6 +741,137 @@ void testCapacity(String maxBatchSize, String maxBatchLingerMs) throws IOExcepti
+
+            allTasksTimeMs = TopicOperatorPerformanceUtils.processAllTopicsConcurrently(testStorage, numberOfTasks, spareEvents, warmUpTasksToProcess);
+
+            // Calculate total execution time in nanoseconds and then convert to ms
```
Move to line 843.

fvaleri reviewed (commented):
We are pretty close to what I measured for 0.41, thanks. 

Just one more change. We don't need the additional metrics collection for this test, as we are only interested in the e2e reconciliation time. Wdyt?

see-quick commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -737,6 +741,137 @@ void testCapacity(String maxBatchSize, String maxBatchLingerMs) throws IOExcepti
+        // this means we would invoke create/update/delete operations x times in one iteration
+        int[] batchEventSizes = IntStream.iterate(seed, n -> n + seed).limit(limit).toArray();
+        return Stream.of(
+            new Object[][]{
+                    {"100", "100"},     // Default configuration
+                    {"100", "10"},      // Default batch size, with lower linger time
+                    {"10", "1"},        // Minimal batching for high responsiveness
+                    {"50", "100"},      // Moderate batching for balanced performance
+                    {"100", "500"},     // Heavier batching for throughput focus
+                    {"500", "1000"},    // Extreme batching to test upper limits of performance
+                }
```
2 cases are enough I think

see-quick commented:
> We are pretty close to what I measured for 0.41, thanks.
> 
> Just one more change. We don't need the additional metrics collection for this test, as we are only interested in the e2e reconciliation time. Wdyt?

Agree, I was just curious what values would all metrics have during observation 👍

see-quick commented:
@strimzi-ci run tests --cluster-type=ocp --cluster-version=4.15 --install-type=bundle --profile=performance --testcase=TopicOperatorPerformance#testSystemScalability

fvaleri commented:
LGTM, but I'm getting this error at the end:

```sh
2024-06-14 12:42:49 [main] INFO  [BasePerformanceMetricsParser:311] Using user.dir: /home/fvaleri/Documents/strimzi/strimzi-kafka-operator/to-perf/systemtest
Error handling metrics: /home/fvaleri/Documents/strimzi/strimzi-kafka-operator/to-perf/systemtest/target/performance
java.nio.file.NoSuchFileException: /home/fvaleri/Documents/strimzi/strimzi-kafka-operator/to-perf/systemtest/target/performance
```

fvaleri reviewed (approved):
LGTM. Thanks.

henryZrncik commented on systemtest/src/main/java/io/strimzi/systemtest/performance/report/TopicOperatorPerformanceReporter.java:
```diff
@@ -25,13 +26,34 @@ public class TopicOperatorPerformanceReporter extends BasePerformanceReporter {
+            dirPathBuilder.append("-number-of-topics-").append(numberOfTopics);
+        }
+
+        String processType = null;
```
maybe add a comment here what process type exactly means/represent as it is used heavily in following lines but not in very straight forward manner ?

see-quick replied:
Added

henryZrncik reviewed (approved):
LGTM thanks!

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/PerformanceUtils.java:
```diff
@@ -0,0 +1,29 @@
+    // ensuring that object can not be created outside of class
+    private PerformanceUtils() {}
+
+    /**
+     * Converts a List of Long objects to a primitive long array.
+     * @param longList The List<Long> to be converted.
+     * @return A primitive long array containing the elements of the given List.
+     */
+    public static long[] convertListToPrimitiveArray(List<Long> longList) {
+        return longList.stream()
+            .mapToLong(Long::longValue)
+            .toArray();
+    }
```
Is this really something just for the performance tests? How it is related to performance stuff?

see-quick replied:
It seems it's no longer used. But I have used it for a series of measured times in performance tests where I measured time execution, modification time or creation time in multiple iterations....but I will remove this as it is no longer used. Thanks

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/report/TopicOperatorPerformanceReporter.java:
```diff
@@ -25,13 +26,34 @@ public class TopicOperatorPerformanceReporter extends BasePerformanceReporter {
+
+        final Path topicOperatorUseCasePathDir = performanceLogDir.resolve(dirPathBuilder.toString());
+
+        LOGGER.info("Resolved performance log directory: {} for use case '{}'. Max batch size: {}, Max linger time: {}, " +
+                "Clients enabled: {}, Number of topics: {}, Process type: {}",
+            topicOperatorUseCasePathDir, useCaseName, maxBatchSize, maxBatchLingerMs, clientsEnabled,
+            numberOfTopics.equals("0") ? "not included" : numberOfTopics, processType == null ? "not included" : processType);
```
I don't know, should it be on `info` level?
Also, the name of the directory is self explaining, so maybe printing the full path would be sufficient? That way you would not need the `processType` there and checking if it's null or not.

see-quick replied:
> I don't know, should it be on info level?

I can change it to debug...but I don't see any problem with having it to INFO cause it's only printed once... :)

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,299 @@
+    }
+
+    private static void performCreationWithWait(int start, int end, ExtensionContext currentContext, TestStorage testStorage) {
+        ResourceManager.setTestContext(currentContext);
```
Why the test context is set here, isn't it something that RM should handle by itself?

see-quick replied:
Because this method is invoked in the new `Thread`. Therefore if you do not set context you would end up in the NPE because a new Thread does not hold the state of the `ExtensionContext`, and so you need to set it.

im-konge replied:
Okay, I would add a comment about this there, why it's needed. (I know that for you it makes sense, but when someone else will go through the code, it will be similarly confusing for him as it was for me)

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,299 @@
+     * @param end           the ending index of topics to delete
+     */
+    public static void deleteKafkaTopicsInRange(String namespace, String prefix, int start, int end) {
+        List<KafkaTopic> topicsToDelete = KafkaTopicResource.kafkaTopicClient().inNamespace(namespace).list().getItems()
+            .stream()
+            .filter(topic -> {
+                String name = topic.getMetadata().getName();
+                if (!name.startsWith(prefix)) return false;
+                // Extract index from the topic name
+                String indexPart = name.substring(prefix.length());
+                try {
+                    int index = Integer.parseInt(indexPart);
+                    return index >= start && index < end;
+                } catch (NumberFormatException e) {
+                    LOGGER.error("Failed to parse index from Kafka topic name: {}", name, e);
+                    return false;
+                }
+            }).toList();
```
I guess this is something that would be useful also somewhere else? At least I would not bind it just to `performance`

see-quick replied:
Cool, I would move it to KafkaTopicUtils

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,299 @@
+        LOGGER.info("Deleted Kafka topics from {} to {} in namespace {} with prefix {}", start, end, namespace, prefix);
+    }
+
+    public static List<KafkaTopic> getAllKafkaTopicsWithPrefix(String namespace, String prefix, int start, int end) {
```
The name is a bit misleading, as it's not just listing the topics with prefix, but also in the range.
Also, why it's not used in `deleteKafkaTopicsInRange`?

see-quick replied:
Yeah, leftover....when I copy-paste this method from somewhere...fixed with in range suffix. Thanks for it :)

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,299 @@
+
+        long allTasksTimeMs = (System.nanoTime() - startTime) / 1_000_000;
+
+        if (warmUpTasksToProcess != 0) {
+            // boundary between tests => less likelihood that tests would influence each other
+            LOGGER.info("Cooling down");
+            try {
+                Thread.sleep(30_000);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
```
It's fine that you are giving some space between the tests, but I'm not sure I understand how they can influence each other. You have deployed one Kafka cluster and you are just deploying the KafkaTopics with different configurations, so UTO would have some kind of issue with that? The magical number there is not helping much as well.
And `Cooling down` is not much saying TBH

see-quick replied:
So as @fvaleri suggested with a smaller checking interval would it make sense? I am also not sure how much it helps cause mine is a little bit different from yours (i.e., you have a shared Kafka cluster where you do creation/modification and deletion) but in this case, we deploy a brand new Kafka cluster. So if that still makes sense I can make it to 10 seconds or 5 seconds.

im-konge replied:
If it's needed, fine let's have some lower duration of this sleep. But I would definitely change the log message.

see-quick replied:
Updated to 5 seconds

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
You named it `TopicOperatorPerformanceUtils`, but it seems that you are working a lot of with threading. Should it be some kind of object or different handler rather than "utils"?

see-quick replied:
I am not sure...there not so much going on in that class. Only starting executor :) If you think it needs a separate handling class I could create that.

im-konge replied:
I think that "utils" makes it a bit weird. Because of the stuff you are doing in there. But maybe it's just me, WDYT @strimzi/system-test-contributors

see-quick replied:
^^ @strimzi/system-test-contributors :)

fvaleri commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -737,6 +741,112 @@ void testCapacity(String maxBatchSize, String maxBatchLingerMs) throws IOExcepti
         }
     }
 
+    private static Stream<Arguments> provideConfigurationsForTestSystemScalability() {
+        final int seed = 50;
+        final int limit = 20; // up to 1000 topics to create
+
+        // this means we would invoke create/update/delete operations x times in one iteration
+        int[] batchEventSizes = IntStream.iterate(seed, n -> n + seed).limit(limit).toArray();
+        return Stream.of(
+            new Object[][]{
+                    {"100", "100"},     // Default configuration
+                    {"100", "10"}       // Default batch size, with lower linger time
+                }
+            )
+            .flatMap(config -> Arrays.stream(batchEventSizes)
+                .mapToObj(numberOfTopics -> {
+                    int eventPerTask = 3;
+                    int numberOfTasks = numberOfTopics / eventPerTask;
+                    int spareEvents = numberOfTopics % eventPerTask;
+
+                    return Arguments.of(config[0], config[1], numberOfTasks, spareEvents);
+                }));
+    }
```
That's a good idea. IMO, the full test is something to run on demand, mainly to check some big change to the Topic Operator. This is something every maintainer or contributor can do while developing or reviewing, not just me :)

im-konge replied:
Yeah well, I agree that this should do every maintainer or contributor. But I know how it ends actually. I think it would be more user-friendly if these kind of tests could be more configurable. It's also better for debugging the issues, where you would just pass the config to the test and would not have to exclude the testcases or waiting a long time before you come to the point which interests you.

That way you will be able to remove the cool downs and other stuff that are present in the suite.

fvaleri replied:
I think cooling down is not bad, but we can reduce to 10 or 5 seconds.

Frawless replied:
I agree with Lukas. Why do we have this pipeline that seems to be quite complex, AFAIU it is not easy to trigger just a specific configuration but you have to run all of the tests and wait...it doesn't sounds great. We have troubles for executing acceptance tests or regression on TF so having one huge pipeline with this execution time is not much usable from my POV. 

Why we don't have something a little bit shorter? Like scenario where you will create 500 topics and check it will be finished in some reasonable time? Or do we already have this kind of baseline that could be executed as part of regression? Kafka has it's benchmarks tests and AFAIR these doesn't take 10+hours.

fvaleri replied:
The objective here is to test the TO scalability, so you need to scale with increasing number of topic events. This test doesn't take that long when running locally, but I see your concerns when running on CI pipeline with less resources. In order to make it faster we can maybe run a shorter sequence of batches (i.e. 200, 400, 800, 1000). This takes around 2 minutes on my laptop. Would that work? I'm also not sure on which pipeline this would run and how it can be triggered.

Frawless replied:
I guess even 1-2 hours are fine, but when I saw the pipeline running in Jenkins last week, it timeouted after 22h which is crazy. The timeout on TF is not set to 30h which is crazy as well. If it takes only 2 minutes, why we need 30h timeout? Regarding resources....30GB RAM and 8CPU is not so small I would say.

fvaleri replied:
> If it takes only 2 minutes, why we need 30h timeout?

It's around two minutes with the changes I proposed in my previous comment.

I actually didn't review the pipeline config as I don't know it very well, but I guess it was tuned for the initial implementation a needs to be revised now. 

@see-quick knows more for sure.

see-quick replied:
In the previous commits, we have like 7 different configurations and each configuration tried to go from 50 to 1000 events (increasing per 50). Now we have only 2 configurations. :) So execution time is now not a problem. 

> As I mentioned offline, it's fine that you are doing these things, but what about having some smaller configuration that will be run for every pipeline and then some possibility to users that will run the tests (IMHO it will be mainly @fvaleri) to actually configure the stuff with what the tests should be run.
With this, it will run for a long time and in case that the test fails, you will, again, have to re-run all the tests? That doesn't apply just for this test, but for all other perf tests.

Okay, so we can decrease everything to pick only two configurations... 👍 as we have now this scalability test WDYT?

see-quick commented on systemtest/tmt/tests/strimzi/main.fmf:
```diff
@@ -93,4 +93,14 @@ adjust:
+/performance-topic-operator-user-workload:
+  summary: Run topic operator performance user workload strimzi test suite
+  tags: [strimzi, kafka, kraft, performance]
+  duration: 30h
```
Yeah, as I already stated above this is not needed I think 12h or even less could be okay.

see-quick commented on systemtest/src/main/java/io/strimzi/systemtest/utils/kafkaUtils/KafkaTopicScalabilityUtils.java:
```diff
@@ -86,21 +86,26 @@ public static void waitForTopicsReady(String namespaceName, String topicPrefix,
-            }));
+        for (int i = startIndex; i < endIndex; i++) {
+            String currentTopic = topicPrefix + "-" + i;
+            KafkaTopicUtils.waitForTopicConfigContains(namespaceName, currentTopic, config);
```
It was not so much faster as you would expect as I have tested it...There were some hidden a bug when something went wrong resulting in prolonging execution time (one would expect that when one WaitException arises) it would mean that everything will stop but in that case other threads don't know about that so they continue to spawning more topics....until KafkaTopic will not be ready. So I have switched to sequential here because of better debugging, plus such methods are then used in concurrent execution in mine new created methods.

im-konge replied:
Okay, makes sense, thanks.

fvaleri commented:
@im-konge @Frawless I played a bit with this work from @see-quick and this is the result:

https://github.com/fvaleri/strimzi-kafka-operator/blob/to-perf/systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java

It should make it faster and easier to maintain, while still providing value in assessing the impact on performance of TO changes. Let me know what you think. Thanks.

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,375 @@
+     * @param topicName          The topic name.
+     * @param currentContext     The current context of the test case.
+     */
+    public static void deleteTopicWithWait(String namespace, String topicName, ExtensionContext currentContext) {
```
You don't have to pass the `ExtensionContext` anywhere or? You can do the `ResourceManager.getTestContext()`

see-quick replied:
This is leftover (i.e., no longer used) cause I have to re-work it. Removing this code but thanks for pointing it out.

im-konge commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,375 @@
+    }
+
+    private static void updateTopic(String namespace, String topicName, KafkaTopicSpec topicSpec) {
+        LOGGER.debug("Updating KafkaTopic {}/{}", namespace, topicName);
+        KafkaTopicResource.replaceTopicResourceInSpecificNamespace(topicName, kafkaTopic -> kafkaTopic.setSpec(topicSpec), namespace);
+    }
+
+    public static void waitForTopicConfigContains(String namespace, String topicName, Map<String, Object> config) {
+        LOGGER.debug("Waiting for KafkaTopic: {}/{} to contain correct config", namespace, topicName);
+        TestUtils.waitFor("KafkaTopic: " + namespace + "/" + topicName + " to contain correct config",
+                TestConstants.GLOBAL_POLL_INTERVAL, TestConstants.GLOBAL_STATUS_TIMEOUT,
+                () -> KafkaTopicUtils.configsAreEqual(KafkaTopicResource.kafkaTopicClient()
+                        .inNamespace(namespace).withName(topicName).get().getSpec().getConfig(), config)
+        );
+        LOGGER.debug("KafkaTopic: {}/{} contains correct config", namespace, topicName);
+    }
+
+    private static void deleteKafkaTopic(String namespace, String topicName) {
+        LOGGER.debug("Deleting topic {}", topicName);
+        ResourceManager.cmdKubeClient().namespace(namespace).deleteByName(KafkaTopic.RESOURCE_SINGULAR, topicName);
+    }
+
+    public static void waitForKafkaTopicDeletion(String namespaceName, String topicName) {
+        LOGGER.debug("Waiting for KafkaTopic: {}/{} deletion", namespaceName, topicName);
+        TestUtils.waitFor("deletion of KafkaTopic: " + namespaceName + "/" + topicName,
+                TestConstants.POLL_INTERVAL_FOR_RESOURCE_READINESS,
+                TestConstants.GLOBAL_TIMEOUT_SHORT,
+                () -> {
+                    if (KafkaTopicResource.kafkaTopicClient().inNamespace(namespaceName).withName(topicName).get() == null) {
+                        return true;
+                    } else {
+                        LOGGER.warn("KafkaTopic: {}/{} is not deleted yet! Triggering force delete by cmd client!", namespaceName, topicName);
+                        cmdKubeClient(namespaceName).deleteByName(KafkaTopic.RESOURCE_KIND, topicName);
+                        return false;
+                    }
+                },
+                () -> LOGGER.info(KafkaTopicResource.kafkaTopicClient().inNamespace(namespaceName).withName(topicName).get())
+        );
+    }
+}
```
Aren't some of the methods already implemented in the `KafkaTopicUtils` or something like that? I would move it outside of the "performance utils" as they can be (IMHO) used somewhere else

see-quick replied:
It was part of Federico's aux methods but I then realised that they are not needed cause everything can be replaced by
```java
/**
     * Executes the full lifecycle of Topic Operator tasks which includes creation, modification,
     * and deletion operations. These operations are encapsulated as a single task that is suitable
     * for parallel processing.
     *
     * @param topicIndex        the index of the topic which identifies the specific topic to be managed.
     * @param testStorage       an object representing the storage where test data or states are maintained.
     * @param extensionContext  an object representing current context of the test case
     */
    private static void performFullLifecycle(int topicIndex, TestStorage testStorage, ExtensionContext extensionContext) {
        performCreationWithWait(topicIndex, topicIndex + 1, extensionContext, testStorage);
        performModificationWithWait(topicIndex, topicIndex + 1, extensionContext, testStorage, KAFKA_TOPIC_CONFIG_TO_MODIFY);
        performDeletionWithWait(topicIndex, topicIndex + 1, extensionContext, testStorage);
    }
```
and so I have replaced it with that. So this method will be not here.

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -91,25 +93,11 @@ private static Stream<Arguments> provideConfigurationsForAliceBulkBatchUseCase()
-                Arguments.of("100", "30000", false),  // Lower batch size with 30 seconds linger time
-                Arguments.of("500", "30000", false),  // Increased batch size with 30 seconds linger time
-                Arguments.of("1000", "30000", false), // Large batch size with 30 seconds linger time
+//                Arguments.of("100", "10", false),     // Lower batch size with short linger time for comparison
```
This should be here?

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -91,25 +93,11 @@ private static Stream<Arguments> provideConfigurationsForAliceBulkBatchUseCase()
-                Arguments.of("100", "100", true),    // Lower batch size with longer linger time for extended comparison
-                Arguments.of("500", "100", true),    // Increased batch size with longer linger time
-                Arguments.of("1000", "100", true)    // Large batch size with longer linger time
+//                Arguments.of("100", "10", true)     // Lower batch size with short linger time for comparison
```
This should be here?

see-quick replied:
removed

im-konge commented on systemtest/src/test/java/io/strimzi/systemtest/performance/TopicOperatorPerformance.java:
```diff
@@ -303,10 +291,8 @@ public void testAliceBulkBatchUseCase(String maxBatchSize, String maxBatchLinger
-            Arguments.of("250", "5000", false), // Medium batch size with longer no-op timed reconciliation intervals for reduced CPU utilization
-            Arguments.of("500", "5000", false)  // Larger batch size with longer intervals, balancing scalability with operational latency
+            Arguments.of("250", "1000", false) // Medium batch size with 1 second linger time, optimized for infrequent changes
+//            Arguments.of("500", "1000", false)  // Slightly larger batch size with 1 second linger time, considering memory usage efficiency
```
This should be here?

see-quick replied:
removed

im-konge reviewed (commented):
Few comments, thanks for the changes

see-quick commented:
/packit test --labels performance

Frawless commented on .packit.yaml:
```diff
@@ -312,3 +312,24 @@ jobs:
         tmt:
           name: regression-operands
+  - job: tests
+    trigger: release
+    identifier: "performance"
+    targets:
+      - centos-stream-9-x86_64
+      - centos-stream-9-aarch64
+    skip_build: true
+    manual_trigger: true
```
You can't have `trigger: release` and `manual_trigger: true` together. it will never get executed. Also you won't be able to run any on PRs, just on releases and I don't think it's a good idea.

see-quick replied:
Updated ;)

see-quick commented:
/packit test --labels performance

see-quick commented:
/packit test --labels performance

see-quick commented:
/packit test --labels performance

see-quick commented:
/packit test --labels performance

im-konge reviewed (approved):
LGTM, thanks

Frawless commented on systemtest/tmt/plans/main.fmf:
```diff
@@ -236,22 +236,8 @@ finish:
-
-/performance-topic-operator-capacity:
-  summary: Run topic operator performance capacity strimzi test suite
+/capacity:
```
do we still need baremetal machines for performance plan? And for capacity?

see-quick replied:
Good point. I think we don't need it anymore for performance. But for capacity, I think yes.

henryZrncik reviewed (approved):
LGTM, thanks for PR!

Frawless reviewed (approved):
Just several nits

see-quick commented on systemtest/src/main/java/io/strimzi/systemtest/performance/PerformanceConstants.java:
```diff
@@ -176,6 +186,6 @@ public interface PerformanceConstants {
      * Performance specific related tags
      */
-    String PERFORMANCE_CAPACITY = "performance-capacity";
+    String CAPACITY = "capacity";
```
> Guess this should be next to the others tags in case it is used as `@Tag()`. I dont say you have to move it in this PR, but. maybe it might be reasonable to create something like `Tags.java` with tags related constatns. Or just moves it to `Constants`

Noted. I can do it in the the next PR :) Dedicated class for Tags.

see-quick commented on systemtest/src/main/java/io/strimzi/systemtest/performance/report/TopicOperatorPerformanceReporter.java:
```diff
@@ -25,13 +26,36 @@ public class TopicOperatorPerformanceReporter extends BasePerformanceReporter {
+
+        final Path topicOperatorUseCasePathDir = performanceLogDir.resolve(dirPathBuilder.toString());
+
+        LOGGER.info("Resolved performance log directory: {} for use case '{}'. Max batch size: {}, Max linger time: {}, " +
```
updated

see-quick commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,264 @@
+        ResourceManager.setTestContext(currentContext);
+        LOGGER.info("Creating Kafka topics from index {} to {}", start, end);
+        KafkaTopicScalabilityUtils.createTopicsViaK8s(testStorage.getNamespaceName(), testStorage.getClusterName(),
+            testStorage.getTopicName(), start, end, 12, 3, 2);
```
Updated ;) created variables for it.

see-quick commented on systemtest/src/main/java/io/strimzi/systemtest/performance/utils/TopicOperatorPerformanceUtils.java:
```diff
@@ -0,0 +1,264 @@
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();
+        LOGGER.info("All topic lifecycles completed.");
+
+        long allTasksTimeMs = (System.nanoTime() - startTime) / 1_000_000;
```
Yeah, good catch thanks, updated to
```java
long allTasksTimeMs = Duration.ofNanos(System.nanoTime() - startTime).toMillis();
```

see-quick commented:
/packit test --labels performance
